stjude/proteinpaint
# detect and handle race conditions in the matrix and hiercluster

## Description

Fixes https://gdc-ctds.atlassian.net/browse/SV-2373.

The issue is due to late-arriving server data requests that could overwrite more recent data requests from the latest state update. The solution is to track the action sequence ID that triggered a state update + server request, and if that ID has changed by the time the server response is complete, then do not process that stale data.

To test:
- http://localhost:3000/testrun.html?dir=plots&name=hierCluster.integration
- http://localhost:3000/testrun.html?dir=mass&name=matrix.integration, may change the last test to `tape.only`

To cause the test to FAIL, comment out line 238 of `rx/index.js`, to verify that the implemented solution works to prevent a simulated race condition. It would not be readily apparent that the matrix has a wrong row label, until the matrix `cell.__data__` is inspected showing the data.value `key` does not match the gene label. This sneaky error is quite bad because nothing looks out of the ordinary.

## Checklist

[Check each task](https://github.com/stjude/proteinpaint/wiki/Pull-Request-Checklist) that has been performed or verified to be not applicable.
